### ISSUE TITLE
Consider analyze-graph for packages and producing graph?

### ISSUE THREAD

**Issue by NeilGirdhar** [astral-sh/ruff#13431]
My [Graph-Imports](https://github.com/NeilGirdhar/graph_imports/tree/main) package has been a useful tool in my development.  Now that Ruff has `ruff analyze graph`, it could be extended to supplant my tool if there's a desire.  Suggestions:

- Add a `--package` flag to switch from producing a graph of module imports to package imports (take the union over every module within a package?)
- Add a `--output_image` flag to produce an actual image of the graph rather than text?

I realize this may be out of scope, but I thought I'd suggest it since it might be easy to do.  (The source for graph-imports is quite short: https://github.com/NeilGirdhar/graph_imports/blob/main/graph_imports/__init__.py)

**Comment by charliermarsh** [comment]
The first idea looks great. The second I have to think on a little more... It'd be nice to output in a format that visualization tools can support, like [Mermaid](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams)? But that's just my immediate reaction :)

**Comment by NeilGirdhar** [comment]
One last suggestion:  The main reason I even have this tool is to detect cycles.  I think that even cycles between package imports reveals a possible design error.  So, reporting cycles here may also be something to consider.

PS Love all the new tools you guys are making by the way.  UV has totally revolutionized my workflow with how fast it is compared with poetry.

**Comment by charliermarsh** [comment]
Oh yeah cycles is very cool. The infrastructure we built here also supports tracing third-party dependencies so we can add that eventually too which is fun.

**Comment by purajit** [comment]
Big +1! I've implemented cycle detection as well (which I assume will eventually be part of ruff's port of R0401 once the design for file traversal is decided), and also a drop-in replacement for import-linter, and it requires some faffing to get the module paths. So a +1 to third-party deps inclusion as well!

**Comment by nickdrozd** [comment]
Pyreverse produces import graphs in various formats (Graphviz, Mermaid, etc). It would be useful to have something like that.

https://pylint.readthedocs.io/en/latest/additional_tools/pyreverse/index.html

**Comment by tmr232** [comment]
If producing a graphviz output, clustering by package hierarchy would also be nice, and improve readability.

**Comment by rgon** [comment]
I have implemented graph generation in python here, it's similar to the `packages` output of `pyreverse`:
https://gist.github.com/rgon/115595265b1d5566f5734eb04bf97736

It has to be run with a pipe, to fetch the data from `ruff analyze`, although since ruff is super fast, it's feasible to run it on CI without much manual labor:
```sh
ruff analyze graph | python3 ./depgraph.py -c -fd
```
It also adds @tmr232 's suggestion to cluster by package and also by submodule, since it's quite more legible that way for larger projects!

It does not have:
+ class reports like `pyreverse`
+ cycle detection/highlighting

Example output:
![Image](https://github.com/user-attachments/assets/9965cb70-1e90-40e8-b5ce-4124046d2a8e)

**Comment by seddonym** [comment]
Hi! Just came across this thread. I'm the maintainer of [Import Linter](https://import-linter.readthedocs.io/), [Grimp](https://grimp.readthedocs.io/) (the Rust-based tool, used by Import Linter, that builds the graph of Python imports in the first place), and [Impulse](https://impulse-cli.readthedocs.io/) (visualizes dependencies within Python projects).

We're currently working on [a new contract type for Import Linter that detects cycles](https://github.com/seddonym/import-linter/pull/279), may be of interest.

I'm not sure what plans Astral has in this space, but do reach out if you see any opportunities for collaboration!
